src/hotspot/share/opto/escape.cpp

```diff
@@ -3634,47 +4050,75 @@ void ConnectionGraph::split_unique_types(GrowableArray<Node *>  &alloc_worklist,
 
   // Update the memory inputs of MemNodes with the value we computed
   // in Phase 2 and move stores memory users to corresponding memory slices.
-  // Disable memory split verification code until the fix for 6984348.
-  // Currently it produces false negative results since it does not cover all cases.
-#if 0 // ifdef ASSERT
-  visited.Reset();
-  Node_Stack old_mems(arena, _compile->unique() >> 2);
-#endif
   for (uint i = 0; i < ideal_nodes.size(); i++) {
     Node*    n = ideal_nodes.at(i);
     Node* nmem = get_map(n->_idx);
     assert(nmem != NULL, "sanity");
     if (n->is_Mem()) {
-#if 0 // ifdef ASSERT
-      Node* old_mem = n->in(MemNode::Memory);
-      if (!visited.test_set(old_mem->_idx)) {
-        old_mems.push(old_mem, old_mem->outcnt());
-      }
-#endif
       assert(n->in(MemNode::Memory) != nmem, "sanity");
       if (!n->is_Load()) {
         // Move memory users of a store first.
-        move_inst_mem(n, orig_phis);
+        move_inst_mem(n, &orig_phis);
       }
       // Now update memory input
       igvn->hash_delete(n);
       n->set_req(MemNode::Memory, nmem);
       igvn->hash_insert(n);
       record_for_optimizer(n);
     } else {
-      assert(n->is_Allocate() || n->is_CheckCastPP() ||
+      assert(n->is_Allocate() || n->is_ReducedAllocationMerge() || n->is_CheckCastPP() ||
              n->is_AddP() || n->is_Phi(), "unknown node used for set_map()");
     }
   }
-#if 0 // ifdef ASSERT
-  // Verify that memory was split correctly
-  while (old_mems.is_nonempty()) {
-    Node* old_mem = old_mems.node();
-    uint  old_cnt = old_mems.index();
-    old_mems.pop();
-    assert(old_cnt == old_mem->outcnt(), "old mem could be lost");
+
+  // Update the memory inputs of ReducedAllocationMerge nodes
+  Unique_Node_List ram_nodes;
```

Comment:
Same here: pass over the whole graph to enumerate `ReducedAllocationMerge` nodes. For such cases, we keep node lists on Compile (see `GrowableArray<Node*>` declared there).